### PR TITLE
Refactor xmp parsing

### LINKED ISSUE #48: Switch regex with xmptodict
Regex sucks

### PR DESCRIPTION
**Pull request by elliott-imhoff**
# Refactor xmp parsing
## What?
Switch from using regex for xmp parsing to using the `xmltodict` library. Now we don't have to deal with the horrible complexities of trying to create a regular expression for every new tag we want to parse. Now all the tags are accessed via dictionary keys, the exact same way as exif tags which is nice. Xmp tags are now stored in `xmp.py` with classes representing the tags for various sensor types, similar to the classes storing the regex patterns we had before. We can easily grab the xmp tags for a sensor by calling `xmp.get_tags()`. If a sensor doesn't have a specific xmp tag supported, it will throw a `KeyError` the same as if a supported tag wasn't found so we can handle all parsing errors the same.

Other changes:
* I made image path required for every function and updated the docstrings to make it clear users shouldn't pass `exif_data` or `xmp_data` to the functions. Since we have had memoization, parsing the exif/xmp data separately and passing it in for a speed bump has been unnecessary for a while. 
* Updated the errors raised. `FileNotFoundError` is thrown if an image path doesn't exist instead of a `ValueError`. `ParsingError` thrown otherwise in nearly all expected error cases.
* `alt_source` for relative altitude can't be "session" anymore. "session" is just a last case fallback for old Sentera imagery, and we should be reading the relative altitude tag in all cases where it is present. Additionally, renamed the laser range finder string for `alt_source` to "lrf" instead of "rlf" which seemed to be a typo.
* All non-Sentera timestamps are made timezone aware using the image lat/lon (before this was for just DJI imagery). Sentera timestamps are already timezone aware.
* `get_ils()` had the `use_clear_channel` flag removed. Seems this was just used for experimentation.
* Updated tests, removing a lot of unnecessary checks. Added new tests for `get_bandnames()` and `get_wavelength_data()`

## PR Checklist
- [x] Merged latest master
- [x] Updated version number
- [x] Version numbers match between package ``__version__`` and *pyproject.toml*
- [x] All private git packages are at their newest version in both *pyproject.toml* and *environment.yml*
- [x] All git package version numbers match between *pyproject.toml* and *environment.yml*
## Breaking Changes
* image_path is now required for all functions
* `alt_source` in `get_relative_altitude()` renamed to "lrf" instead of "rlf" for laser range finder
* `use_clear_channel` flag removed from `get_ils()`
## Notes
Closes #48

### CODE REVIEW

**Review by SamCWill — COMMENTED** [review]
Looks great! Love these changes. I'm still not sure about the removal of the ILS Clear channel getter. This is what py-radiometric-corrections currently uses for all D4K images. I think input from Zach Marston is required. I've sent him a private message just incase he overlooked the group one

File: imgparse/imgparse.py [hunk 1]
```diff
@@ -212,74 +197,98 @@ def get_camera_params(
...
+        raise ParsingError("Couldn't parse session altitude from session.txt")
+
+    return float(session_alt)
+
+
 @get_if_needed("exif_data", getter=get_exif_data, getter_args=["image_path"])
 @get_if_needed("xmp_data", getter=get_xmp_data, getter_args=["image_path"])
 def get_relative_altitude(
-    image_path, exif_data=None, xmp_data=None, alt_source="relative"
+    image_path, exif_data=None, xmp_data=None, alt_source="default"
```
[the remark this answers is not in the document]
Comment (elliott-imhoff) [reply imgparse/imgparse.py:236]: In a following PR, I'm adding a new `alt_source` that specifies to use the terrain api

File: imgparse/imgparse.py [hunk 2]
```diff
@@ -536,41 +484,69 @@ def get_gsd(
...
 @get_if_needed("xmp_data", getter=get_xmp_data, getter_args=["image_path"])
-def get_bandnames(image_path=None, xmp_data=None):
+def get_bandnames(image_path=None, exif_data=None, xmp_data=None):
     """
     Get the name of each band of an image.
 
-    :param image_path: the full path to the image (optional if `xmp_data` provided)
-    :param xmp_data: the XMP data of image, as a string dump of the original XML (optional to speed up processing)
+    :param image_path: the full path to the image
+    :param exif_data: used internally for memoization. Not necessary to supply.
```
[the remark this answers is not in the document]
Comment (elliott-imhoff) [reply imgparse/imgparse.py:541]: If you've already parsed any other data from the image, the exif data is already cached, so no delay. Then I just timed the `get_exif_data()` and `get_xmp_data()` functions for 10000 calls after removing memoization. `get_exif_data()` is 0.002 sec per call. `get_xmp_data()` is 0.0003 sec per call.

File: imgparse/getters.py [hunk 3]
```diff
@@ -18,21 +26,28 @@ def get_xmp_data(image_path):
...
+    if isinstance(xmp_dict, list):
+        temp_dict = {}
+        for d in xmp_dict:
+            temp_dict.update(d)
+        xmp_dict = temp_dict
+    # Remove '@' signs, which appear to be non-consistent
+    for k in list(xmp_dict):
+        if k[0] == "@":
+            xmp_dict[k[1:]] = xmp_dict.pop(k)
+    return xmp_dict
```
Comment (cliff-dellamano) [inline imgparse/getters.py:50]:
```suggestion
xmp_dict = {k: v.lstrip('@') for k, v in xmp_dict.items()}
```

File: imgparse/getters.py [hunk 4]
```diff
@@ -18,21 +26,28 @@ def get_xmp_data(image_path):
...
+    # If there are too many xmp tags, returned as list
+    if isinstance(xmp_dict, list):
+        temp_dict = {}
+        for d in xmp_dict:
+            temp_dict.update(d)
+        xmp_dict = temp_dict
+    # Remove '@' signs, which appear to be non-consistent
+    for k in list(xmp_dict):
+        if k[0] == "@":
+            xmp_dict[k[1:]] = xmp_dict.pop(k)
```
Comment (cliff-dellamano) [inline imgparse/getters.py:49]:
```suggestion
xmp_dict = {k: v.lstrip('@') for k, v in xmp_dict.items()}
```

File: imgparse/imgparse.py [hunk 5]
```diff
@@ -142,64 +129,62 @@ def get_timestamp(image_path=None, exif_data=None, format_string="%Y:%m:%d %H:%M
...
+    make, model = get_make_and_model(image_path, exif_data)
     if make == "Sentera":
         datetime_obj = pytz.utc.localize(datetime_obj)
-    elif make == "DJI":
-        lat, lon = get_lat_lon(image_path=image_path, exif_data=exif_data)
+    else:
+        lat, lon = get_lat_lon(image_path, exif_data)
         timezone = pytz.timezone(TimezoneFinder().timezone_at(lng=lon, lat=lat))
         datetime_obj = timezone.localize(datetime_obj)
-    else:
```
Comment (cliff-dellamano) [inline imgparse/imgparse.py:158]: Why is it safe to remove this case?

**Review by cliff-dellamano — APPROVED** [review]
LGTM, just a few comments

File: imgparse/imgparse.py (hunk 5 above)
Comment (elliott-imhoff) [reply imgparse/imgparse.py:158]: We can localize datetime by lat/lon for all non-Sentera sensor makes, not just DJI

Comment (elliott-imhoff) [reply imgparse/imgparse.py:158]: See here: https://github.com/SenteraLLC/py-image-metadata-parser/pull/49/files/29b146deb076676d51d00717d07dcda4041607bc#diff-69bd3a008c817f4c9feda3865ab8473e91a99e6c6026f908531c4eb38e35ad62R143

**Pull request merged**